src/profile-logic/symbolication.js

```diff
@@ -391,133 +419,190 @@ export function applySymbolicationStep(
   const {
     frameTable: oldFrameTable,
     funcTable: oldFuncTable,
+    nativeSymbols: oldNativeSymbols,
     stringTable,
   } = thread;
   const { threadLibSymbolicationInfo, resultsForLib } = symbolicationStepInfo;
   const {
     resourceIndex,
     allFramesForThisLib,
     allFuncsForThisLib,
+    allNativeSymbolsForThisLib,
+    libIndex,
   } = threadLibSymbolicationInfo;
 
   const availableFuncs: Set<IndexIntoFuncTable> = new Set(allFuncsForThisLib);
-  const frameToFuncAddressMap: Map<IndexIntoFrameTable, Address> = new Map();
-  const funcAddressToSymbolNameMap: Map<Address, string> = new Map();
-  const funcAddressToCanonicalFuncIndexMap: Map<
+  const availableNativeSymbols: Set<IndexIntoFuncTable> = new Set(
+    allNativeSymbolsForThisLib
+  );
+  const frameToSymbolAddressMap: Map<IndexIntoFrameTable, Address> = new Map();
+  const symbolAddressToNameMap: Map<Address, string> = new Map();
+  const symbolAddressToCanonicalSymbolIndexMap: Map<
     Address,
-    IndexIntoFuncTable
+    IndexIntoNativeSymbolTable
   > = new Map();
 
-  // We want to group frames into funcs, and give each func a name.
-  // We group frames to the same func if the addresses for these frames resolve
-  // to the same funcAddress.
+  // We want to group frames into nativeSymbols, and give each nativeSymbol a name.
+  // We group frames to the same nativeSymbol if the addresses for these frames resolve
+  // to the same symbolAddress.
   // We obtain the funcAddress from the symbolication information in resultsForLib:
   // resultsForLib does not only contain the name of the function; it also contains,
   // for each address, the symbolAddress.
-  // All frames with the same symbolAddress are grouped into the same function.
-
+  // All frames with the same symbolAddress are grouped into the same nativeSymbol.
+  // Afterwards, we create funcs for symbols with the same name, and then group frames
+  // into funcs.
   for (const frameIndex of allFramesForThisLib) {
-    const oldFrameFunc = oldFrameTable.func[frameIndex];
+    const oldFrameSymbol = oldFrameTable.nativeSymbol[frameIndex];
     const address = oldFrameTable.address[frameIndex];
     let addressResult: AddressResult | void = resultsForLib.get(address);
     if (addressResult === undefined) {
-      const oldSymbol = stringTable.getString(oldFuncTable.name[oldFrameFunc]);
-      addressResult = {
-        symbolAddress: oldFuncTable.address[oldFrameFunc],
-        name: oldSymbol,
-      };
+      if (oldFrameSymbol !== null) {
```

Comment:
Was there a reason why we weren't checking this before? Was it guaranteed to be non-null?

Reply:
That's right, the types guarantees the func to be non-null. All frames always have a func, because the call tree is based on funcs. But with native symbols, the native symbol of a frame is null before symbolication.